File: go/merge_sort/merge_sort_ints.go
package merge_sort

// MergeSortInts should sort int array with merge sort
func MergeSortInts(arr []int) {
	mergeSortInts(arr, make([]int, len(arr)))
}

func mergeSortInts(arr, tmp []int) {
	n := len(arr)
	m := n / 2
	if m == 0 {
		return
	}

	mergeSortInts(arr[:m], tmp)
	mergeSortInts(arr[m:], tmp)

	i, j, k := 0, m, 0
	for ; i < m || j < n; k++ {
		if j == n || i < m && arr[i] <= arr[j] {
			tmp[k] = arr[i]
			i++
		} else {
			tmp[k] = arr[j]
			j++
		}
	}

	copy(arr, tmp)
}

File: python/merge_sort/merge_sort_ints.py

# merge_sort_ints should sort int array with merge sort
def merge_sort_ints(arr: list[int]):
	# raise NotImplementedError('TODO')

	merge_sort_ints_recurse(arr, arr.copy(), 0, len(arr))


def merge_sort_ints_recurse(arr, tmp, l, r):
	if r-l <= 1:
		return
	m = (l+r)//2
	merge_sort_ints_recurse(arr, tmp, l, m)
	merge_sort_ints_recurse(arr, tmp, m, r)

	merge(arr, tmp, l, m, r)


def merge(arr, tmp, l, m, r):
	i, j, k = l, m, 0
	while i < m or j < r:
		if i < m and (j == r or arr[i] <= arr[j]):
			tmp[k] = arr[i]
			i += 1
		else:
			tmp[k] = arr[j]
			j += 1
		k += 1
	arr[l:r] = tmp[:k]
